src/jupyterlabcontroller/services/builder/lab.py

```diff
@@ -0,0 +1,821 @@
+        lab
+            Specification of the lab requested.
+        image
+            Image to use for the lab.
+        secrets
+            Dictionary of secrets to expose to the lab.
+        pull_secret
+            Optional pull secret for the lab pod.
+
+        Returns
+        -------
+        LabObjects
+            Kubernetes objects that make up the user's lab.
+        """
+        return LabObjects(
+            namespace=self._build_namespace(user.username),
+            config_maps=self._build_config_maps(user, lab, image),
+            network_policy=self._build_network_policy(user.username),
+            pvcs=self._build_pvcs(user.username),
+            quota=self._build_quota(user),
+            secrets=self._build_secrets(user.username, secrets, pull_secret),
+            service=self._build_service(user.username),
+            pod=self._build_pod(user, lab, image),
+        )
+
+    def build_lab_config_volumes(
+        self,
+        username: str,
+        volumes: list[LabVolume],
+        prefix: str = "",
+    ) -> list[LabVolumeContainer]:
+        """Construct volumes and mounts for volumes in the lab configuration.
+
+        Parameters
+        ----------
+        username
+            Name of user this lab is for, used to name the PVC objects.
+        volumes
+            Configured volumes. This is a parameter rather than being taken
+            straight from the lab configuration because this method is also
+            used when constructing init containers, which use separate volume
+            lists.
+        prefix
+            Prefix to prepend to all mount paths, if given.
+
+        Returns
+        -------
+        list of LabVolumeContainer
+            List of volumes and mounts.
+        """
+        results = []
+        pvc_count = 1
+        for storage in volumes:
+            is_readonly = storage.mode == FileMode.RO
+            name = storage.container_path.replace("/", "-")[1:].lower()
+            match storage.source:
+                case HostPathVolumeSource() as source:
+                    volume = V1Volume(
+                        name=name,
+                        host_path=V1HostPathVolumeSource(path=source.path),
+                    )
+                case NFSVolumeSource() as source:
+                    volume = V1Volume(
+                        name=name,
+                        nfs=V1NFSVolumeSource(
+                            path=source.server_path,
+                            read_only=is_readonly,
+                            server=source.server,
+                        ),
+                    )
+                case PVCVolumeSource():
+                    claim = V1PersistentVolumeClaimVolumeSource(
+                        claim_name=f"{username}-nb-pvc-{pvc_count}",
+                        read_only=is_readonly,
+                    )
+                    volume = V1Volume(name=name, persistent_volume_claim=claim)
+                    pvc_count += 1
```

Comment:
It feels so weird without a default case!

Reply:
Type checking is great for that!